libsql-sqlite3/src/wal.c

```diff
@@ -4250,150 +4151,210 @@ static void libsqlGetWalPathname(char *buf, const char *orig, int orig_len) {
-  if (rc != SQLITE_OK) {
-    return NULL;
-  }
-#endif
+int sqlite3LogExists(void* self, sqlite3_vfs *vfs, const char *main_db_path_name, int *exists) {
+    char *zWal;
+    int rc = libsqlMakeWalPathname(main_db_path_name, &zWal);
+    if (rc != 0) return rc;
+    rc = sqlite3OsAccess(vfs, zWal, SQLITE_ACCESS_EXISTS, exists);
+    sqlite3_free(zWal);
+    if (rc != 0) return rc;
+    return SQLITE_OK;
+}
 
-  if (!zName || *zName == '\0') {
-    zName = "default";
-  }
-  if (methods_head == NULL && strncmp(zName, "default", 7) == 0) {
-    methods.iVersion = 1;
-    methods.xOpen = sqlite3WalOpen;
-    methods.xClose = sqlite3WalClose;
-    methods.xLimit = sqlite3WalLimit;
-    methods.xBeginReadTransaction = sqlite3WalBeginReadTransaction;
-    methods.xEndReadTransaction = sqlite3WalEndReadTransaction;
-    methods.xFindFrame = sqlite3WalFindFrame;
-    methods.xReadFrame = sqlite3WalReadFrame;
-    methods.xDbsize = sqlite3WalDbsize;
-    methods.xBeginWriteTransaction = sqlite3WalBeginWriteTransaction;
-    methods.xEndWriteTransaction = sqlite3WalEndWriteTransaction;
-    methods.xUndo = sqlite3WalUndo;
-    methods.xSavepoint = sqlite3WalSavepoint;
-    methods.xSavepointUndo = sqlite3WalSavepointUndo;
-    methods.xFrames = sqlite3WalFrames;
-    methods.xCheckpoint = sqlite3WalCheckpoint;
-    methods.xCallback = sqlite3WalCallback;
-    methods.xExclusiveMode = sqlite3WalExclusiveMode;
-    methods.xHeapMemory = sqlite3WalHeapMemory;
+int sqlite3LogDestroy(void* self, sqlite3_vfs *vfs, const char *main_db_path_name) {
+    char *zWal;
+    int rc = libsqlMakeWalPathname(main_db_path_name, &zWal);
+    if (rc != 0) return rc;
+    rc = sqlite3OsDelete(vfs, zWal, 0);
+    sqlite3_free(zWal);
+    if (rc != 0) return rc;
+    return SQLITE_OK;
+}
 
-#ifdef SQLITE_ENABLE_SNAPSHOT
-    methods.xSnapshotGet = sqlite3WalSnapshotGet;
-    methods.xSnapshotOpen = sqlite3WalSnapshotOpen;
-    methods.xSnapshotRecover = sqlite3WalSnapshotRecover;
-    methods.xSnapshotCheck = sqlite3WalSnapshotCheck;
-    methods.xSnapshotUnlock = sqlite3WalSnapshotUnlock;
-#endif
+/*
+** Open a connection to the WAL file zWalName. The database file must
+** already be opened on connection pDbFd. The buffer that zWalName points
+** to must remain valid for the lifetime of the returned Wal* handle.
+**
+** Custom virtual methods may be provided via the pMethods parameter.
+**
+** A SHARED lock should be held on the database file when this function
+** is called. The purpose of this SHARED lock is to prevent any other
+** client from unlinking the WAL or wal-index file. If another process
+** were to do this just after this client opened one of these files, the
+** system would be badly broken.
+**
+** If the log file is successfully opened, SQLITE_OK is returned and
+** *ppWal is set to point to a new WAL handle. If an error occurs,
+** an SQLite error code is returned and *ppWal is left unmodified.
+*/
+static int sqlite3WalOpen(
+  void *self,
```

Comment:
I don't follow why this pointer is type-erased, but we must at least name it anyway. Similarly, `sqlite3_file` object is never really instantiated, but instead cast to a specific type before it's used, e.g. a unix file. I can't suggest a name for it before I understand what `self` is here, isn't it just an implementation of a specific WAL, so something like `libsql_wal_impl *`? This type can be public and still be opaque, same as `sqlite3_file`, you just declare it in one of the public headers as
```
typedef struct libsql_wal_impl libsql_wal_impl;
```

We can't afford `void *` here because this type name also serves as documentation

Reply:
self is the method receiver. It's whatever state the WAL needs to function. In the case of sqlite3, this is a sqlite3_wal.

I don't get the analogy with `sqlite3_file`? the `sqlite3_file` has a pointer to the methods, but `self` can really be anything. whatever is returned by `xOpen`

Reply:
ping: instead of void, it should be a named opaque type, something like `libsql_wal_impl *`, defined as `typedef struct libsql_wal_impl *libsql_wal_impl`.